crates/dojo-world/src/manifest.rs

```diff
@@ -106,133 +83,283 @@ pub struct ComputedValueEntrypoint {
-    pub contracts: Vec<Contract>,
-    pub models: Vec<Model>,
+#[serde(tag = "kind")]
+pub struct Contract {
+    #[serde_as(as = "UfeHex")]
+    pub class_hash: FieldElement,
+    pub abi: Option<String>,
+    #[serde_as(as = "Option<UfeHex>")]
+    pub address: Option<FieldElement>,
 }
 
-impl Manifest {
-    /// Load the manifest from a file at the given path.
-    pub fn load_from_path(path: impl AsRef<Path>) -> Result<Self, std::io::Error> {
-        let file = fs::File::open(path)?;
-        Ok(Self::try_from(file)?)
+impl ManifestMethods for Contract {
+    fn abi(&self) -> Option<&String> {
+        self.abi.as_ref()
     }
 
-    /// Writes the manifest into a file at the given path. Will return error if the file doesn't
-    /// exist.
-    pub fn write_to_path(self, path: impl AsRef<Path>) -> Result<(), std::io::Error> {
-        let fd = fs::File::options().write(true).open(path)?;
-        Ok(serde_json::to_writer_pretty(fd, &self)?)
+    fn set_abi(&mut self, abi: Option<String>) {
+        self.abi = abi;
     }
 
-    /// Construct a manifest of a remote World.
-    ///
-    /// # Arguments
-    /// * `provider` - A Starknet RPC provider.
-    /// * `world_address` - The address of the remote World contract.
-    pub async fn load_from_remote<P>(
+    fn class_hash(&self) -> &FieldElement {
+        self.class_hash.as_ref()
+    }
+
+    fn set_class_hash(&mut self, class_hash: FieldElement) {
+        self.class_hash = class_hash;
+    }
+}
+
+#[serde_as]
+#[derive(Clone, Default, Debug, Serialize, Deserialize, PartialEq)]
+#[serde(tag = "kind")]
+pub struct Class {
+    #[serde_as(as = "UfeHex")]
+    pub class_hash: FieldElement,
+    pub abi: Option<String>,
+}
+
+impl ManifestMethods for Class {
+    fn abi(&self) -> Option<&String> {
+        self.abi.as_ref()
+    }
+
+    fn set_abi(&mut self, abi: Option<String>) {
+        self.abi = abi;
+    }
+
+    fn class_hash(&self) -> &FieldElement {
+        self.class_hash.as_ref()
+    }
+
+    fn set_class_hash(&mut self, class_hash: FieldElement) {
+        self.class_hash = class_hash;
+    }
+}
+
+pub trait ManifestMethods {
+    fn abi(&self) -> Option<&String>;
+    fn set_abi(&mut self, abi: Option<String>);
+    fn class_hash(&self) -> &FieldElement;
+    fn set_class_hash(&mut self, class_hash: FieldElement);
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, Debug)]
+pub struct Manifest<T>
+where
+    T: ManifestMethods,
+{
+    #[serde(flatten)]
+    pub inner: T,
+    pub name: SmolStr,
+}
+
+pub struct BaseManifest {
+    pub world: Manifest<Class>,
+    pub executor: Manifest<Class>,
+    pub base: Manifest<Class>,
+    pub contracts: Vec<Manifest<Class>>,
+    pub models: Vec<Manifest<DojoModel>>,
+}
+pub struct DeployedManifest {
+    pub world: Manifest<Contract>,
+    pub executor: Manifest<Contract>,
+    pub base: Manifest<Class>,
+    pub contracts: Vec<Manifest<DojoContract>>,
+    pub models: Vec<Manifest<DojoModel>>,
+}
+
+impl BaseManifest {
```

Comment:
`BaseManifest` can be confusing with the actual `base` contract, at least it confused me anytime I read the PR hehe.

What we want to express here is the manifest outputted by the compilation, right? Maybe something related to that could be more explicit.

Reply:
haha yes, i used `Base` prefix since it was suggested in the issue. `overlay` didn't fit very well so replaced it with `Deployed`.

> What we want to express here is the manifest outputted by the compilation, right? Maybe something related to that could be more explicit.

true but we ideally want something that understandable by users (i.e. people using dojo to build games) in that case i think compilation is implementation detail.

https://github.com/dojoengine/dojo/pull/1482#discussion_r1476456200